proposals/0000-likelihood-annotations.rst

```diff
@@ -0,0 +1,264 @@
+Add Likelihood Annotations
+==============
+
+.. proposal-number:: _.
+.. trac-ticket:: _.
+.. implemented:: _.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/182>`_.
+.. sectnum::
+.. contents::
+
+A lot of performance can be gained by optimizing for hot code paths.
+However GHC currently does not allow users to actually give hints about which code
+paths are hot code paths.
+
+This proposal changes this.
+
+
+Motivation
+------------
+
+Programmers often know a good deal more about program behaviour than we can currently
+communicate to the compiler. This leads to missed optimization opportunities.
+
+Consider types like ``data Either a b = Left a | Right b``
+
+Left is often used to represent an exceptional case. Making Right the common case.
+
+Based on this information and given code like ``either f g x`` we might want
+to inline g but not f for example.
+
+The same logic applies to many other cases. For example Lists are rarely empty.
+
+A programmer has this information but no way to communicate this to the compiler.
+
+Communicating this kind of information can allow optimization passes to produce
+faster code. In particular, potential gains include:
+ * Avoid inlining into unlikely case alternatives.
+ * Produce better code layout.
+ * Better register allocation.
+
+This would also bring us closer to profile guided optimization (PGO).
+A lot of the required backend work for this is also required for PGO.
+
+Proposed Change Specification
+-----------------------------
+
+We propose a new Pragma: ``{-# LIKELY <NUM> #-}``
+
+This will be useable in the following constructs:
+ - Simple case expressions.
+ - Simple function bindings.
+ - If expressions.
+ - Data type declarations for Sum Types.
+
+Uses where the requirements are not satisfied will result in warnings similar to
+the ones from bad UNPACK pragmas. Likelihood values must be >= 0.
+
+Weight semantics
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+Given a simple branching expression with n branches ``[A1 .. An]``,
+with likelihoods ``[L1 .. Ln]`` GHC will optimize code under the assumption that
+the chance for the i-th branch to be taken is ``Li / sum [L1 .. Ln].``
```

Comment:
How would you decide on the likelihoods when one branch is more likely and the other is "normally" likely? 2 and 1? 10 and 1? 100 and 1? Perhaps it would be good to allow some branches to not be annotated and choose some default likelihood for them?

Reply:
I chose the meaning of weights to be relative execution frequencies. The downside to this is that the only proper way to know what weights to use is by knowing what "normally" likely and "more" likely is. Either through profiling or domain knowledge.

> Perhaps it would be good to allow some branches to not be annotated and choose some default likelihood for them?

I don't think so. If the code is performance sensitive enough for one to care about the weights it seems justified to expect weights for all branches. If the code isn't performance sensitive enough then it's likely better not to use weights at all.

For what it's worth the proposal talks about this further down. In particular not annotating a branch:

* Will trigger a warning if any of the other branches is annotated.
* The compiler is then free to choose any weight for this branch.

The reason why I believe this should be a warning is that I don't think the compiler can properly infer a "default" weight. If the other branches have weights [1,1,0,x] there is no easy way for the compiler to know what unknown branches should default to. The highest weight? The average? The median? The lowest weight? Lowest non-zero? All of these seem reasonable in some situations. I would probably have the compiler chose the median. But what if the compiler can tell the median is likely a bad choice by some further analysis?

Overall I think it's just simpler *and* better to require full weights and leave unannotated branches unspecified/up to the implementation at this stage.